File: 1-algo-toolbox/2-algorithmic-warmup/8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/fibonacci_sum_squares.cpp
#include <iostream>
#include <vector>

using namespace std;

long long fibonacci_sum_squares(long long n) {
    vector<long long> fibList;

    long long a = 0;
    long long b = 1;

    int sum = 1;

    for (int i = 2; i <= n; i++)
    {
        long long temp = b;
        b = (a + b) % 10;
        a = temp;

        long long sq_digit = (b * b) % 10;
        
        sum = (sum + sq_digit) % 10;
    }

    return sum;
}

int fibonacci_sum_squares_naive(long long n) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;
    long long sum      = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
        sum += current * current;
    }

    return sum % 10;
}

int main() {
    long long n = 0;
    std::cin >> n;
    std::cout << fibonacci_sum_squares(n) << endl;
}

File: 1-algo-toolbox/2-algorithmic-warmup/8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/fibonacci_sum_squares.py
# Uses python3

def fibonacci_sum_squares(n):

    if n <= 1:
        return n
    
    previous = 0
    current = 1

    for _ in range(n):
        previous, current = current, (previous + current) 
    
    return current*previous % 10

def fibonacci_sum_squares_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1
    sum      = 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        sum += current * current

    return sum % 10


n = int(input())
print(fibonacci_sum_squares(n%30))

# for i in range(2000):
#     print(i)
#     if fibonacci_sum_squares(i%30) != fibonacci_sum_squares_naive(i):
#         break